http-gateways/IPLD_GATEWAY.md

```diff
@@ -0,0 +1,196 @@
+# IPLD Gateway Specification
+
+**Authors**:
+
+- Mauve Signweaver ([@RangerMauve](https://github.com/RangerMauve))
+
+----
+
+**Abstract**
+
+IPLD Gateway is an extension of [PATH_GATEWAY.md](./PATH_GATEWAY.md) that enables lower level interaction with IPLD data structures under a specific path.
+
+This document describes the delta between [PATH_GATEWAY.md](./PATH_GATEWAY.md) and this gateway type.
+
+Summary:
+
+- Adds a new `/ipld/{cid}[/{segments}][?{params}]` subpath to the gateway
+- Defines a specification for parsing out extra parameters for individual path segments.
+- Describes how to map to and from `ipld://` URLs
+
+## HTTP API
+
+### `GET /ipld/{cid}[/{segments}][?{params}]`
+
+Resolve IPLD paths to some data.
+
+The path segments will be traversed with any parameters used to transform data along the way.
+
+The `format` query string parameter, or the `Accept` request header can be used to control the format which will be used to return the data.
+
+By default, data will be returned as `dag-json`
+
+<!--
+TODO: Cache control semantics?
+TODO: Add more details on how the traversal works?
+-->
+
+### `HEAD /ipld/{cid}[/{segments}][?{params}]`
+
+Resolves IPLD paths, and yields the same status code and headers as `GET`.
+
+### `POST /ipld/` and `POST /ipld/localhost/`
+
+Upload raw data to IPLD.
+The `body` of the request shall be parsed according to the `Content-Type` as IPLD data via standard encodings.
+`/localhost/` is used to support `POST ipld://localhost/` for uploading IPLD data to local nodes in web browsers that support it.
+
+The response will contain an `ipld://{cid}/` URL pointing at your data.
```

Comment:
Spec should remove any ambiguity:
- Contain it where? (A) plain text in response  body?  B) a `Location` header? 
- What will be content-type of the response? `text/plain` ?

Reply:
This is something I'd like to clarify with @fabricedesre since we had a bit of a disagreement.

Right now the precendent within Kubo and Agregore's protocol handlers is that there will be a 201 response with a `Location` header containing the URL as well as an empty body.

Fabrice was into having a 200 response and the URL inside the response body, which is something I was originally doing in Agregore, but switch when we started extending the writable gateway functionality in Kubo.

Ideally we should settle on the best course of action here during Lisbon. 😅

Reply:
Ideally, I'd like to use this to inform all the other protocol handlers too.